io_scene_niftools/modules/nif_export/geometry/mesh/__init__.py

```diff
@@ -349,11 +367,17 @@ def export_tri_shapes(self, b_obj, n_parent, trishape_name=None):
             # update tangent space (as binary extra data only for Oblivion)
             # for extra shader texture games, only export it if those textures are actually exported
             # (civ4 seems to be consistent with not using tangent space on non shadered nifs)
-            if mesh_uv_layers and mesh_hasnormals:
-                if bpy.context.scene.niftools_scene.game in ('OBLIVION', 'FALLOUT_3', 'SKYRIM') or (bpy.context.scene.niftools_scene.game in self.texture_helper.USED_EXTRA_SHADER_TEXTURES):
-                    if bpy.context.scene.niftools_scene.game == 'SKYRIM':
-                        tridata.bs_num_uv_sets = tridata.bs_num_uv_sets + 4096
-                    trishape.update_tangent_space(as_extra=(bpy.context.scene.niftools_scene.game == 'OBLIVION'))
+            if use_tangents:
+                if game == 'SKYRIM':
+                    tridata.bs_num_uv_sets = tridata.bs_num_uv_sets + 4096
+                # calculate the bitangents using the normals, tangent list and bitangent sign
+                bitlist = bitangent_sign_list * np.cross(normlist, tanlist)
+                # B_tan: +d(B_u), B_bit: +d(B_v) and N_tan: +d(N_v), N_bit: +d(N_u)
+                # moreover, N_v = 1 - B_v, so d(B_v) = - d(N_v), therefore N_tan = -B_bit and N_bit = B_tan
+                self.add_defined_tangents(trishape,
+                                          tangents=-bitlist,
```

Comment:
This is confusing. So you're exporting blender's bitangents as nif tangents? Seems like axis correction weirdness.

Do we get a performance gain from calculating the bitangents with numpy over crossing blender vectors?

Reply:
Correct, it is axis correction weirdness. In Nifs, tangents point to the positive (Nif) V direction and bitangents point to the positive (Nif) U direction (hence `N_tan: +d(N_v), N_bit: +d(N_u)`), while Blender tangents point in the positive U direction, and bitangents in the positive V direction. That's why they're flipped (tangents become bitangents and vice versa).

Then, the `Nif V coordinate = 1 - Blender V coordinate`, so the direction of positive V is flipped (hence the minus sign).
I tried to leave that information in the comments as concisely as possible, but maybe it should be in the technical documentation somewhere, too.

As for the speed of calculating the bitangents with numpy vs individually, I haven't tested the alternative. However, one of numpy's big strengths is array operations, so I would be extremely surprised if it were any slower than calculating them individually.

Reply:
> This is confusing. So you're exporting blender's bitangents as nif tangents? Seems like axis correction weirdness.
> 
> Do we get a performance gain from calculating the bitangents with numpy over crossing blender vectors?

Where would be the best place in the documentation to outline why tangent space is calculated from Blender the way that it is? I would assume it doesn't need to be in user, but there doesn't really seem to be a good place in it in development.

Reply:
We can create a design decisions section for anything that we feel needs enough of an explaination about it.